File: mRemoteNG/UI/UI.Window.PortScan.vb
﻿Imports mRemoteNG.Tools.PortScan
Imports mRemoteNG.My
Imports WeifenLuo.WinFormsUI.Docking
Imports mRemoteNG.App.Runtime

Namespace UI
    Namespace Window
        Public Class PortScan
            Inherits Base
#Region "Constructors"
            Public Sub New(ByVal panel As DockContent, ByVal import As Boolean)
                InitializeComponent()

                WindowType = Type.PortScan
                DockPnl = panel
                _import = import
            End Sub
#End Region

#Region "Private Properties"
            Private ReadOnly Property IpsValid() As Boolean
                Get
                    If String.IsNullOrEmpty(ipStart.Octet1) Then Return False
                    If String.IsNullOrEmpty(ipStart.Octet2) Then Return False
                    If String.IsNullOrEmpty(ipStart.Octet3) Then Return False
                    If String.IsNullOrEmpty(ipStart.Octet4) Then Return False

                    If String.IsNullOrEmpty(ipEnd.Octet1) Then Return False
                    If String.IsNullOrEmpty(ipEnd.Octet2) Then Return False
                    If String.IsNullOrEmpty(ipEnd.Octet3) Then Return False
                    If String.IsNullOrEmpty(ipEnd.Octet4) Then Return False

                    Return True
                End Get
            End Property
#End Region

#Region "Private Fields"
            Private ReadOnly _import As Boolean
            Private _portScanner As Scanner
            Private _scanning As Boolean = False
#End Region

#Region "Private Methods"
#Region "Event Handlers"
            Private Sub PortScan_Load(ByVal sender As System.Object, ByVal e As EventArgs) Handles MyBase.Load
                ApplyLanguage()

                Try
                    If _import Then
                        lvHosts.Columns.AddRange(New ColumnHeader() {clmHost, clmSSH, clmTelnet, clmHTTP, clmHTTPS, clmRlogin, clmRDP, clmVNC})
                        ShowImportControls(True)
                        cbProtocol.SelectedIndex = 0
                    Else
                        lvHosts.Columns.AddRange(New ColumnHeader() {clmHost, clmOpenPorts, clmClosedPorts})
                        ShowImportControls(False)
                    End If
                Catch ex As Exception
                    MessageCollector.AddExceptionMessage(My.Language.strPortScanCouldNotLoadPanel, ex)
                End Try
            End Sub

            Private Sub portStart_Enter(sender As System.Object, e As EventArgs) Handles portStart.Enter
                portStart.Select(0, portStart.Text.Length)
            End Sub

            Private Sub portEnd_Enter(sender As System.Object, e As EventArgs) Handles portEnd.Enter
                portEnd.Select(0, portEnd.Text.Length)
            End Sub

            Private Sub btnScan_Click(ByVal sender As System.Object, ByVal e As EventArgs) Handles btnScan.Click
                If _scanning Then
                    StopScan()
                Else
                    If IpsValid Then
                        StartScan()
                    Else
                        MessageCollector.AddMessage(Messages.MessageClass.WarningMsg, My.Language.strCannotStartPortScan)
                    End If
                End If
            End Sub

            Private Sub btnImport_Click(ByVal sender As System.Object, ByVal e As EventArgs) Handles btnImport.Click
                Dim protocol As mRemoteNG.Connection.Protocol.Protocols = Tools.Misc.StringToEnum(GetType(mRemoteNG.Connection.Protocol.Protocols), cbProtocol.SelectedItem)

                Dim hosts As New List(Of ScanHost)
                For Each item As ListViewItem In lvHosts.SelectedItems
                    Dim scanHost As ScanHost = TryCast(item.Tag, ScanHost)
                    If scanHost IsNot Nothing Then hosts.Add(item.Tag)
                Next

                App.Import.ImportFromPortScan(hosts, protocol)

                DialogResult = DialogResult.OK
                Close()
            End Sub
#End Region

            Private Sub ApplyLanguage()
                lblStartIP.Text = String.Format("{0}:", My.Language.strStartIP)
                lblEndIP.Text = String.Format("{0}:", My.Language.strEndIP)
                btnScan.Text = My.Language.strButtonScan
                btnImport.Text = My.Language.strButtonImport
                lblOnlyImport.Text = String.Format("{0}:", My.Language.strProtocolToImport)
                clmHost.Text = My.Language.strColumnHostnameIP
                clmOpenPorts.Text = My.Language.strOpenPorts
                clmClosedPorts.Text = My.Language.strClosedPorts
                Label2.Text = String.Format("{0}:", My.Language.strEndPort)
                Label1.Text = String.Format("{0}:", My.Language.strStartPort)
                TabText = My.Language.strMenuPortScan
                Text = My.Language.strMenuPortScan
            End Sub

            Private Sub ShowImportControls(ByVal controlsVisible As Boolean)
                pnlPorts.Visible = controlsVisible
                pnlImport.Visible = controlsVisible
                If controlsVisible Then
                    lvHosts.Height = pnlImport.Top - lvHosts.Top
                Else
                    lvHosts.Height = pnlImport.Bottom - lvHosts.Top
                End If
            End Sub

            Private Sub StartScan()
                Try
                    _scanning = True
                    SwitchButtonText()
                    lvHosts.Items.Clear()

                    Dim ipAddressStart As Net.IPAddress = Net.IPAddress.Parse(ipStart.Text)
                    Dim ipAddressEnd As Net.IPAddress = Net.IPAddress.Parse(ipEnd.Text)

                    If _import Then
                        _portScanner = New Scanner(ipAddressStart, ipAddressEnd)
                    Else
                        _portScanner = New Scanner(ipAddressStart, ipAddressEnd, portStart.Value, portEnd.Value)
                    End If

                    AddHandler _portScanner.BeginHostScan, AddressOf PortScanner_BeginHostScan
                    AddHandler _portScanner.HostScanned, AddressOf PortScanner_HostScanned
                    AddHandler _portScanner.ScanComplete, AddressOf PortScanner_ScanComplete

                    _portScanner.StartScan()
                Catch ex As Exception
                    MessageCollector.AddMessage(Messages.MessageClass.ErrorMsg, "StartScan failed (UI.Window.PortScan)" & vbNewLine & ex.Message, True)
                End Try
            End Sub

            Private Sub StopScan()
                If _portScanner IsNot Nothing Then _portScanner.StopScan()
                _scanning = False
                SwitchButtonText()
            End Sub

            Private Sub SwitchButtonText()
                If _scanning Then
                    btnScan.Text = My.Language.strButtonStop
                Else
                    btnScan.Text = My.Language.strButtonScan
                End If

                prgBar.Maximum = 100
                prgBar.Value = 0
            End Sub

            Private Shared Sub PortScanner_BeginHostScan(ByVal host As String)
                MessageCollector.AddMessage(Messages.MessageClass.InformationMsg, "Scanning " & host, True)
            End Sub

            Private Delegate Sub PortScannerHostScannedDelegate(ByVal host As ScanHost, ByVal scannedCount As Integer, ByVal totalCount As Integer)
            Private Sub PortScanner_HostScanned(ByVal host As ScanHost, ByVal scannedCount As Integer, ByVal totalCount As Integer)
                If InvokeRequired Then
                    Invoke(New PortScannerHostScannedDelegate(AddressOf PortScanner_HostScanned), New Object() {host, scannedCount, totalCount})
                    Return
                End If

                MessageCollector.AddMessage(Messages.MessageClass.InformationMsg, "Host scanned " & host.HostIp, True)

                Dim listViewItem As ListViewItem = host.ToListViewItem(_import)
                If listViewItem IsNot Nothing Then
                    lvHosts.Items.Add(listViewItem)
                    listViewItem.EnsureVisible()
                End If

                prgBar.Maximum = totalCount
                prgBar.Value = scannedCount
            End Sub

            Private Delegate Sub PortScannerScanComplete(ByVal hosts As List(Of ScanHost))
            Private Sub PortScanner_ScanComplete(ByVal hosts As List(Of ScanHost))
                If InvokeRequired Then
                    Invoke(New PortScannerScanComplete(AddressOf PortScanner_ScanComplete), New Object() {hosts})
                    Return
                End If

                MessageCollector.AddMessage(Messages.MessageClass.InformationMsg, Language.strPortScanComplete)

                _scanning = False
                SwitchButtonText()
            End Sub
#End Region
        End Class
    End Namespace
End Namespace

File: mRemoteV1.ConvertedToC#/UI/UI.Window.PortScan.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Drawing;
using System.Diagnostics;
using System.Windows.Forms;
using mRemoteNG.Tools.PortScan;
using mRemoteNG.My;
using WeifenLuo.WinFormsUI.Docking;
using mRemoteNG.App.Runtime;

namespace mRemoteNG.UI
{
	namespace Window
	{
		public partial class PortScan : Base
		{
			#region "Constructors"
			public PortScan(DockContent panel, bool import)
			{
				Load += PortScan_Load;
				InitializeComponent();

				WindowType = Type.PortScan;
				DockPnl = panel;
				_import = import;
			}
			#endregion

			#region "Private Properties"
			private bool IpsValid {
				get {
					if (string.IsNullOrEmpty(ipStart.Octet1))
						return false;
					if (string.IsNullOrEmpty(ipStart.Octet2))
						return false;
					if (string.IsNullOrEmpty(ipStart.Octet3))
						return false;
					if (string.IsNullOrEmpty(ipStart.Octet4))
						return false;

					if (string.IsNullOrEmpty(ipEnd.Octet1))
						return false;
					if (string.IsNullOrEmpty(ipEnd.Octet2))
						return false;
					if (string.IsNullOrEmpty(ipEnd.Octet3))
						return false;
					if (string.IsNullOrEmpty(ipEnd.Octet4))
						return false;

					return true;
				}
			}
			#endregion

			#region "Private Fields"
			private readonly bool _import;
			private Scanner _portScanner;
				#endregion
			private bool _scanning = false;

			#region "Private Methods"
			#region "Event Handlers"
			private void PortScan_Load(System.Object sender, EventArgs e)
			{
				ApplyLanguage();

				try {
					if (_import) {
						lvHosts.Columns.AddRange(new ColumnHeader[] {
							clmHost,
							clmSSH,
							clmTelnet,
							clmHTTP,
							clmHTTPS,
							clmRlogin,
							clmRDP,
							clmVNC
						});
						ShowImportControls(true);
						cbProtocol.SelectedIndex = 0;
					} else {
						lvHosts.Columns.AddRange(new ColumnHeader[] {
							clmHost,
							clmOpenPorts,
							clmClosedPorts
						});
						ShowImportControls(false);
					}
				} catch (Exception ex) {
					mRemoteNG.App.Runtime.MessageCollector.AddExceptionMessage(mRemoteNG.My.Language.strPortScanCouldNotLoadPanel, ex);
				}
			}

			private void portStart_Enter(System.Object sender, EventArgs e)
			{
				portStart.Select(0, portStart.Text.Length);
			}

			private void portEnd_Enter(System.Object sender, EventArgs e)
			{
				portEnd.Select(0, portEnd.Text.Length);
			}

			private void btnScan_Click(System.Object sender, EventArgs e)
			{
				if (_scanning) {
					StopScan();
				} else {
					if (IpsValid) {
						StartScan();
					} else {
						mRemoteNG.App.Runtime.MessageCollector.AddMessage(mRemoteNG.Messages.MessageClass.WarningMsg, mRemoteNG.My.Language.strCannotStartPortScan);
					}
				}
			}

			private void btnImport_Click(System.Object sender, EventArgs e)
			{
				mRemoteNG.Connection.Protocol.Protocols protocol = mRemoteNG.Tools.Misc.StringToEnum(typeof(mRemoteNG.Connection.Protocol.Protocols), cbProtocol.SelectedItem);

				List<ScanHost> hosts = new List<ScanHost>();
				foreach (ListViewItem item in lvHosts.SelectedItems) {
					ScanHost scanHost = item.Tag as ScanHost;
					if (scanHost != null)
						hosts.Add(item.Tag);
				}

				mRemoteNG.App.Import.ImportFromPortScan(hosts, protocol);

				DialogResult = DialogResult.OK;
				Close();
			}
			#endregion

			private void ApplyLanguage()
			{
				lblStartIP.Text = string.Format("{0}:", mRemoteNG.My.Language.strStartIP);
				lblEndIP.Text = string.Format("{0}:", mRemoteNG.My.Language.strEndIP);
				btnScan.Text = mRemoteNG.My.Language.strButtonScan;
				btnImport.Text = mRemoteNG.My.Language.strButtonImport;
				lblOnlyImport.Text = string.Format("{0}:", mRemoteNG.My.Language.strProtocolToImport);
				clmHost.Text = mRemoteNG.My.Language.strColumnHostnameIP;
				clmOpenPorts.Text = mRemoteNG.My.Language.strOpenPorts;
				clmClosedPorts.Text = mRemoteNG.My.Language.strClosedPorts;
				Label2.Text = string.Format("{0}:", mRemoteNG.My.Language.strEndPort);
				Label1.Text = string.Format("{0}:", mRemoteNG.My.Language.strStartPort);
				TabText = mRemoteNG.My.Language.strMenuPortScan;
				Text = mRemoteNG.My.Language.strMenuPortScan;
			}

			private void ShowImportControls(bool controlsVisible)
			{
				pnlPorts.Visible = controlsVisible;
				pnlImport.Visible = controlsVisible;
				if (controlsVisible) {
					lvHosts.Height = pnlImport.Top - lvHosts.Top;
				} else {
					lvHosts.Height = pnlImport.Bottom - lvHosts.Top;
				}
			}

			private void StartScan()
			{
				try {
					_scanning = true;
					SwitchButtonText();
					lvHosts.Items.Clear();

					System.Net.IPAddress ipAddressStart = System.Net.IPAddress.Parse(ipStart.Text);
					System.Net.IPAddress ipAddressEnd = System.Net.IPAddress.Parse(ipEnd.Text);

					if (_import) {
						_portScanner = new Scanner(ipAddressStart, ipAddressEnd);
					} else {
						_portScanner = new Scanner(ipAddressStart, ipAddressEnd, portStart.Value, portEnd.Value);
					}

					_portScanner.BeginHostScan += PortScanner_BeginHostScan;
					_portScanner.HostScanned += PortScanner_HostScanned;
					_portScanner.ScanComplete += PortScanner_ScanComplete;

					_portScanner.StartScan();
				} catch (Exception ex) {
					mRemoteNG.App.Runtime.MessageCollector.AddMessage(mRemoteNG.Messages.MessageClass.ErrorMsg, "StartScan failed (UI.Window.PortScan)" + Constants.vbNewLine + ex.Message, true);
				}
			}

			private void StopScan()
			{
				if (_portScanner != null)
					_portScanner.StopScan();
				_scanning = false;
				SwitchButtonText();
			}

			private void SwitchButtonText()
			{
				if (_scanning) {
					btnScan.Text = mRemoteNG.My.Language.strButtonStop;
				} else {
					btnScan.Text = mRemoteNG.My.Language.strButtonScan;
				}

				prgBar.Maximum = 100;
				prgBar.Value = 0;
			}

			private static void PortScanner_BeginHostScan(string host)
			{
				mRemoteNG.App.Runtime.MessageCollector.AddMessage(mRemoteNG.Messages.MessageClass.InformationMsg, "Scanning " + host, true);
			}

			private delegate void PortScannerHostScannedDelegate(ScanHost host, int scannedCount, int totalCount);
			private void PortScanner_HostScanned(ScanHost host, int scannedCount, int totalCount)
			{
				if (InvokeRequired) {
					Invoke(new PortScannerHostScannedDelegate(PortScanner_HostScanned), new object[] {
						host,
						scannedCount,
						totalCount
					});
					return;
				}

				mRemoteNG.App.Runtime.MessageCollector.AddMessage(mRemoteNG.Messages.MessageClass.InformationMsg, "Host scanned " + host.HostIp, true);

				ListViewItem listViewItem = host.ToListViewItem(_import);
				if (listViewItem != null) {
					lvHosts.Items.Add(listViewItem);
					listViewItem.EnsureVisible();
				}

				prgBar.Maximum = totalCount;
				prgBar.Value = scannedCount;
			}

			private delegate void PortScannerScanComplete(List<ScanHost> hosts);
			private void PortScanner_ScanComplete(List<ScanHost> hosts)
			{
				if (InvokeRequired) {
					Invoke(new PortScannerScanComplete(PortScanner_ScanComplete), new object[] { hosts });
					return;
				}

				mRemoteNG.App.Runtime.MessageCollector.AddMessage(mRemoteNG.Messages.MessageClass.InformationMsg, Language.strPortScanComplete);

				_scanning = false;
				SwitchButtonText();
			}
			#endregion
		}
	}
}
